src/ctelnet.cpp

```diff
@@ -1413,6 +1441,91 @@ void cTelnet::processTelnetCommand(const std::string& command)
     case TN_SB: {
         option = command[2];
 
+        // CHARSET
+        if (option == OPT_CHARSET && enableCHARSET) {
+            QByteArray payload = command.c_str();
+            if (payload.size() < 6) {
+                return;
+            }
+
+            // payload is in the Mud Server's encoding, trim off the Telnet suboption
+            // bytes from beginning (3) and end (2):
+            payload = payload.mid(3, static_cast<int>(payload.size()) - 5);
+
+            QString transcodedMsg;
+
+            if (mpOutOfBandDataIncomingCodec) {
+                // Message is encoded
+                transcodedMsg = mpOutOfBandDataIncomingCodec->toUnicode(payload);
+            } else {
+                // Message is in ASCII (though this can handle Utf-8):
+                transcodedMsg = QString::fromUtf8(payload);
+            }
+
+            // remove \r and \n from the data.
+            transcodedMsg.remove(QChar::CarriageReturn);
+            transcodedMsg.remove(QChar::LineFeed);
+            // replace ANSI escape character with escaped version, to handle improperly passed ANSI codes
+            transcodedMsg.replace(QLatin1String("\u001B"), QLatin1String("\\u001B"));
+
+            // CHARSET support per https://tools.ietf.org/html/rfc2066
+            if (command[3] == CHARSET_REQUEST) {
+                // No translate table support.  Discard. 
+                if (transcodedMsg.startsWith(QStringLiteral("[TTABLE]1"))) {
+                    transcodedMsg.remove(QStringLiteral("[TTABLE]1"));
+                }
+
+                // Second character is the separator.
+                QChar separator = transcodedMsg[1];
+                QStringList characterSetList = transcodedMsg.split(separator);
```

Comment:
The character sets' names are to be handled in a case-insensitive way - so one might need to rethink the use of `QByteArrayList::contains(const QByteArray&)` in line 1485 - as it happens all the values in that are in UPPERCASE. Oh, note that `mAcceptableEncodings` is a `QByteArrayList` rather than a `QStringList` and I've already pointed out that this negotiation is done in pure ASCII so there could be no need to message around with transcoding the incoming data and just store it in a `QByteArray` directly...

Reply:
Although it should be done in ASCII, that doesn't mean servers will not already be sending the bytes encoded as something else, like UTF-8.  Are we sure we should chuck that stuff?

Reply:
Remember - since we are not attempting to hand **[EBDIC](https://en.wikipedia.org/wiki/EBCDIC)** - all the encodings we cover have a common set of ASCII code-points for the lower half of the range - i.e. when the MS Bit is not set.